### ISSUE TITLE
NaN values break Grafana Alerting Rules

### ISSUE THREAD

**Issue by a-melnyk** [grafana/grafana#21953]
**What happened**: [grafana alerting aggregation](https://grafana.com/docs/grafana/latest/alerting/rules/#query-condition-example) function **min()** returns math.MaxFloat64 as min value for series which consist of points with NaN values

**What you expected to happen**: it should return 0

**How to reproduce it (as minimally and precisely as possible)**:

- with grafana UI
    - create a widget with a query where denominator can be equal to 0, example for Prometheus Datasource _something/rate(counter[5m])_
    - create a grafana alert and use min() aggregation function in Conditions
    - press "Test Rule" button
    - "Condition[0]: Query Result" should contain points where all values are "null" (they are "NaN"s actually)
    - "Condition[0]: Eval: true, Metric: {instance="localhost:9090"}, Value: 179769313486231570814527423731704356798070567525844996598917476803157260780028538760589558632766878171540458953514382464234321326889464182768467546703537516986049910576551282076245490090389328944075868508455133942304583236903222948165808559332123348274797826204144723168738177180919299881250404026184124858368.000"
    - please see [the dashboard ](https://gist.github.com/a-melnyk/5e48bc63c10164c3ea9d18f079cdfb56)

- please see [the test](https://github.com/a-melnyk/grafana/commit/360d92f5dc40bf549a372c4cc8ee3851922b2de2)

**Anything else we need to know?**: 
I see 2 ways to fix this:

- treat NaN values as invalid values (similarly to nil) in null.FloatFrom(), please see [the commit](https://github.com/a-melnyk/grafana/commit/918a9a6517ba22f39b41293cc16450a6da122acd)
- add an extra check in min() reducer, this seems to be hacky. Please see [the commit](https://github.com/a-melnyk/grafana/commit/ed61925a10163ef99589c3575d2fb69ca7e2fc20)

**Environment**:
- Grafana version: 6.5.2, 6.6.0, master
- Data source type & version: Prometheus 2.8.0, 2.15.2
- OS Grafana is installed on: official docker image
- Grafana plugins: none

**Comment by aknuds1** [comment]
We think this is a duplicate of #13055, but your [fix](https://github.com/a-melnyk/grafana/commit/918a9a6517ba22f39b41293cc16450a6da122acd) looks potentially useful. @kylebrandt @marefr @bergquist what do you think?

**Comment by aknuds1** [comment]
We've decided after all to instead close #13055 and continue work in this issue.

**Comment by a-melnyk** [comment]
Hi Arve, thanks your prompt reply. Should I prepare a Pull Request? Which solution is preferable?

**Comment by aknuds1** [comment]
> Hi Arve, thanks your prompt reply. Should I prepare a Pull Request? Which solution is preferable?

Waiting on decision from @kylebrandt @marefr @bergquist.

**Comment by kylebrandt** [comment]
1. I do not think that converting NaN to "invalid" would be the correct thing to do in the null package. This is because NaN is generally a perfectly valid float outside of JSON.
2. Regarding alert reducers, I think we should just treat NaN as invalid, and exclude them from the reduction for now. Else (at least logically), we would need to handle NaN as the end value of alert evaluation which I don't think is expected.

For next gen alerting I think the reducers should have options on what to do with nan/null/Inf etc.

Side note: wonder what currently happens with `Inf+/-` float values.

**Comment by aknuds1** [comment]
Thanks for chiming in @kylebrandt :)

**Comment by a-melnyk** [comment]
hi @aknuds1 , hi @kylebrandt,

thanks for your answers and explanation. 
Please see the linked PR, I'm eager to receive your feedback.
